File: packaging/test/console/patterns.test.mjs
// Console pattern card tests — zero dependencies. Evaluates lib.js and
// tab-findings.js in one fresh VM context, as the browser loads them.

import { test } from 'node:test';
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import path from 'node:path';
import vm from 'node:vm';
import { fileURLToPath } from 'node:url';

const webDist = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '../../../daemon/internal/api/web_dist');
const ctx = { window: {} };
vm.createContext(ctx);
for (const f of ['lib.js', 'tab-findings.js']) {
  vm.runInContext(readFileSync(path.join(webDist, f), 'utf8'), ctx, { filename: f });
}
const { patternHTML, uncoveredFlags, patternsInView, patternAfterDismiss } = ctx;

const KEY = 'codex|keychain-access|/Users/x/Library/Keychains/login.keychain-db';
const hourly = Array.from({ length: 24 }, (_, i) => (i === 3 ? 323 : 0));
const pattern = (over) => ({
  key: KEY, agent: 'codex', rule: 'keychain-access', title: 'Agent touched the keychain',
  subject: { kind: 'keychain', label: '~/Library/Keychains/login.keychain-db', sub: 'keychain' },
  count: 323, unacked: 323, first: '2026-09-12T03:00:08Z', last: '2026-09-12T03:08:45Z',
  median_gap_s: 1.2, bursts: 320, cadence: 'in bursts a few seconds apart', hourly,
  pids: [40844, 51364], pid_count: 2, sessions: ['s1'], session_count: 1,
  disposition: { state: 'warning', text: 'Needs a look', why: 'Agent touched the keychain' },
  summary: 'codex touched the login keychain 323 times on Sep 12 between 03:00 and 03:08 (2 processes, 1 session), in bursts a few seconds apart.',
  actions: [
    { id: 'mute-class', label: 'Dismiss this flag class', consequence: 'c1', method: 'POST', path: '/mute', body: { rule: 'keychain-access', host: '*' } },
    { id: 'dismiss-all', label: 'Dismiss all 323 open', consequence: 'c2', method: 'POST', path: '/flags/acknowledge', body: { flag_ids: ['k1', 'k2'] } },
    { id: 'kill', label: 'Kill codex (pid 40844)', consequence: 'c3', method: 'POST', path: '/kill', body: { pid: 40844 }, recommended: true },
  ],
  flag_ids: Array.from({ length: 12 }, (_, i) => 'k' + i),
  ...over,
});
const flag = (id) => ({ id, agent: 'codex', rule: 'keychain-access', pid: 40844, ts: '2026-09-12T03:00:08Z', session_id: 's1' });

test('patternHTML: 24 bars, count, summary, open count; actions in served order, recommended first', () => {
  const html = patternHTML(pattern(), Date.parse('2026-09-23T12:00:00Z'), {});
  assert.equal((html.match(/<i class="h\d"><\/i>/g) || []).length, 24);
  assert.match(html, /<i class="h8"><\/i>/);
  assert.match(html, /class="finding pattern-card disp-warning" data-pattern-key="codex\|keychain-access\|\/Users\/x\/Library\/Keychains\/login\.keychain-db"/);
  assert.match(html, /<span class="pattern-meta">323× · /);
  assert.ok(html.includes('codex touched the login keychain 323 times on Sep 12 between 03:00 and 03:08 (2 processes, 1 session), in bursts a few seconds apart.'));
  assert.ok(html.includes('in bursts a few seconds apart · <b class="pattern-open">323 open</b>'));
  assert.ok(html.includes('<p class="finding-verdict">Needs a look: Agent touched the keychain</p>'));
  const ids = [...html.matchAll(/data-action="explain-act" data-pattern-key="[^"]+" data-action-id="([a-z-]+)"/g)].map(m => m[1]);
  assert.deepEqual(ids, ['kill', 'mute-class', 'dismiss-all']);
  assert.match(html, /<button class="btn btn-danger btn-sm" data-action="explain-act"[^>]*>Kill codex<\/button>/);
  assert.ok(!/disabled/.test(html));
  assert.ok(html.includes('<summary>Individual flags (323)</summary>'));
});

test('patternHTML: the card names its processes and launchers, escaped; none when the daemon served none', () => {
  const html = patternHTML(pattern({ processes: [
    { name: 'claude', launcher: 'Claude.app › claude-code 2.1.281', count: 2 },
    { name: '<b>x</b>', count: 1 }] }), Date.parse('2026-09-12T04:00:00Z'), {});
  assert.match(html, /<p class="pattern-processes">claude-code 2\.1\.281 via Claude\.app ×2 · &lt;b&gt;x&lt;\/b&gt; ×1<\/p>/);
  assert.ok(!patternHTML(pattern({}), Date.parse('2026-09-12T04:00:00Z'), {}).includes('pattern-processes'));
});

test('patternHTML: a pattern dismissed in place shows 0 open and disabled buttons', () => {
  const html = patternHTML(pattern({ dismissed: true, unacked: 0, disposition: { state: 'acknowledged', text: 'Reviewed', why: '' } }), Date.now(), {});
  assert.ok(html.includes('<b class="pattern-open">0 open</b>'));
  assert.match(html, /disp-acknowledged/);
  const buttons = html.match(/<button [^>]*data-action="explain-act"[^>]*>/g) || [];
  assert.equal(buttons.length, 3);
  assert.ok(buttons.every(b => b.includes(' disabled')));
});

test('patternHTML: the covered flags behind Details, first 10 then Show more', () => {
  const flags = Array.from({ length: 12 }, (_, i) => flag('k' + i)).concat([flag('other')]);
  const html = patternHTML(pattern(), Date.now(), { flags });
  assert.equal((html.match(/<li>/g) || []).length, 10);
  assert.ok(!html.includes('<code>other</code>'));
  assert.match(html, /data-action="show-more" data-key="pattern:codex\|keychain-access\|[^"]+">Show 2 more</);
  const open = patternHTML(pattern(), Date.now(), { flags, expanded: new Set(['pattern:' + KEY]) });
  assert.equal((open.match(/<li>/g) || []).length, 12);
});

test('uncoveredFlags: a flag a pattern covers is not its own row', () => {
  const rows = uncoveredFlags([flag('k1'), flag('k5'), flag('free')], [pattern()]);
  assert.deepEqual(rows.map(f => f.id), ['free']);
  assert.deepEqual(uncoveredFlags([flag('k1')], []).map(f => f.id), ['k1']);
});

test('patternsInView: search, agent/rule selects and session scope apply to patterns', () => {
  const ps = [pattern(), pattern({ key: 'claude|proxy-secret-leak|api.example.com', agent: 'claude', rule: 'proxy-secret-leak',
    title: 'Secret leaving in agent traffic', subject: { kind: 'connect', label: 'api.example.com' }, summary: 'claude sent a secret', sessions: ['s2'], pids: [9] })];
  const keys = o => patternsInView(ps, o).map(p => p.agent);
  assert.deepEqual(keys({}), ['codex', 'claude']);
  assert.deepEqual(keys({ term: 'api.example' }), ['claude']);
  assert.deepEqual(keys({ term: 'login.keychain' }), ['codex']);
  assert.deepEqual(keys({ agent: 'codex' }), ['codex']);
  assert.deepEqual(keys({ rule: 'proxy-secret-leak' }), ['claude']);
  assert.deepEqual(keys({ session: 's1' }), ['codex']);
  assert.deepEqual(keys({ pids: [9] }), ['claude']);
});

test('patternsInView: a covered flag from a sixth session or pid admits the pattern to that scope; the flag stays hidden', () => {
  const p = pattern({ sessions: ['s1', 's2', 's3', 's4', 's5'], session_count: 6, pids: [1, 2, 3, 4, 5], pid_count: 6 });
  const other = pattern({ key: 'claude|proxy-secret-leak|api.example.com', agent: 'claude', rule: 'proxy-secret-leak',
    sessions: ['s9'], pids: [9], flag_ids: ['c1'] });
  const sixth = { ...flag('k3'), session_id: 's6', pid: 6 };
  const c1 = { ...flag('c1'), agent: 'claude', rule: 'proxy-secret-leak', session_id: 's9', pid: 9 };
  for (const scope of [{ session: 's6' }, { pids: [6] }]) {
    const inView = patternsInView([p, other], { ...scope, flags: [sixth] });
    assert.deepEqual(inView.map(x => x.key), [KEY]);
    assert.deepEqual(uncoveredFlags([sixth], inView).map(f => f.id), []);
    assert.deepEqual(uncoveredFlags([sixth, c1], inView).map(f => f.id), ['c1']);
  }
  assert.deepEqual(patternsInView([p], { session: 's6', flags: [] }).map(x => x.key), []);
  assert.deepEqual(patternsInView([p], { pids: [6], flags: [flag('free')] }).map(x => x.key), []);
});

test('patternAfterDismiss: a capped dismiss-all of 500 ids leaves 2 of 502 open and the card not dismissed', () => {
  const after = patternAfterDismiss(pattern({ count: 502, unacked: 502 }), 500);
  const html = patternHTML(after, Date.now(), {});
  assert.ok(html.includes('<b class="pattern-open">2 open</b>'));
  assert.ok(!after.dismissed);
  assert.ok(!/disabled/.test(html));
  assert.match(html, /disp-warning/);
  const rest = patternAfterDismiss(after, 2);
  assert.equal(rest.dismissed, true);
  assert.ok(patternHTML(rest, Date.now(), {}).includes('<b class="pattern-open">0 open</b>'));
});

test('muteRowHTML: an agent-scoped mute names its agent and carries it on the unmute button', () => {
  const scoped = ctx.muteRowHTML({ rule: 'keychain-access', host: '*', agent: 'codex' });
  assert.ok(scoped.includes('<span class="mute-pair">keychain-access · all hosts · codex</span>'));
  assert.match(scoped, /data-action="unmute" data-rule="keychain-access" data-host="\*" data-agent="codex"/);
  const every = ctx.muteRowHTML({ rule: 'proxy-secret-leak', host: 'api.example.com' });
  assert.ok(every.includes('<span class="mute-pair">proxy-secret-leak · api.example.com</span>'));
  assert.match(every, /data-agent=""/);
  assert.ok(ctx.muteRowHTML({ rule: 'r', host: 'h', agent: '<b>x</b>' }).includes('&lt;b&gt;x&lt;/b&gt;'));
});

test('patternHTML: the served agent-scoped mute label is the button text', () => {
  const actions = pattern().actions.map(a => a.id === 'mute-class'
    ? { ...a, label: 'Mute keychain access for codex', body: { rule: 'keychain-access', host: '*', agent: 'codex' } }
    : a);
  const html = patternHTML(pattern({ actions }), Date.parse('2026-09-23T12:00:00Z'), {});
  assert.match(html, /data-action-id="mute-class"[^>]*>Mute keychain access for codex<\/button>/);
});

File: daemon/internal/api/patterns_test.go
package api

import (
	"encoding/json"
	"fmt"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
	"time"

	"github.com/cavi-ai/secure-agent/daemon/internal/model"
	"github.com/cavi-ai/secure-agent/daemon/internal/resource"
)

const loginKeychain = "/Users/x/Library/Keychains/login.keychain-db"

func keychainFlag(id, agent string, pid int32, ts time.Time, sev int) model.Flag {
	return model.Flag{ID: id, Rule: "keychain-access", Severity: sev, TS: ts, PID: pid, Agent: agent, SessionID: "s1",
		Evidence: []model.EvidenceItem{{Kind: "keychain", Label: loginKeychain, Sub: "keychain access"}}}
}

func onlyPattern(t *testing.T, ps []model.Pattern) model.Pattern {
	t.Helper()
	if len(ps) != 1 {
		t.Fatalf("patterns = %d (%+v), want 1", len(ps), ps)
	}
	return ps[0]
}

func actionByID(acts []model.ExplainAction, id string) *model.ExplainAction {
	if i := actionIndex(acts, id); i >= 0 {
		return &acts[i]
	}
	return nil
}

// A storm of one rule on one subject from two pids is one pattern with its
// cadence and identity.
func TestComputePatternsGroupsBySubject(t *testing.T) {
	a := explainTestAPI(t)
	start := time.Now().Add(-10 * time.Minute)
	for i := 0; i < 330; i++ {
		pid := int32(40844)
		if i >= 320 {
			pid = 51364
		}
		a.store.PutFlag(keychainFlag(fmt.Sprintf("k%03d", i), "codex", pid, start.Add(time.Duration(i)*time.Second), 2))
	}
	p := onlyPattern(t, a.computePatterns(time.Now().Add(-24*time.Hour), 3))
	if p.Count != 330 || p.Unacked != 330 || p.PIDCount != 2 || p.SessionCount != 1 {
		t.Fatalf("count/unacked/pids/sessions = %d/%d/%d/%d, want 330/330/2/1", p.Count, p.Unacked, p.PIDCount, p.SessionCount)
	}
	if p.Bursts < 300 || p.MedianGapS >= 5 {
		t.Fatalf("bursts = %d, median gap = %v; want >= 300 bursts under 5 s", p.Bursts, p.MedianGapS)
	}
	if len(p.PIDs) != 2 || p.PIDs[0] != 40844 {
		t.Fatalf("pids = %v, want the busiest (40844) first", p.PIDs)
	}
	sum := 0
	for _, n := range p.Hourly {
		sum += n
	}
	if sum != 330 || p.Hourly[23] != 330 {
		t.Fatalf("hourly = %v, want 330 in the newest bucket", p.Hourly)
	}
	if p.Key != "codex|keychain-access|"+loginKeychain || p.Subject.Kind != "keychain" || p.Subject.Label != loginKeychain {
		t.Fatalf("key/subject = %q/%+v", p.Key, p.Subject)
	}
	if len(p.FlagIDs) != 330 || p.FlagIDs[0] != "k329" {
		t.Fatalf("flag ids = %d, first %q; want 330, newest first", len(p.FlagIDs), p.FlagIDs[0])
	}
}

// Fewer flags than min are not a pattern; /patterns validates its window.
func TestComputePatternsMinCount(t *testing.T) {
	a := explainTestAPI(t)
	now := time.Now()
	a.store.PutFlag(keychainFlag("a1", "codex", 7, now.Add(-2*time.Minute), 2))
	a.store.PutFlag(keychainFlag("a2", "codex", 7, now.Add(-time.Minute), 2))
	since := now.Add(-time.Hour)
	if ps := a.computePatterns(since, 3); len(ps) != 0 {
		t.Fatalf("min 3 over 2 flags = %+v, want none", ps)
	}
	if p := onlyPattern(t, a.computePatterns(since, 2)); p.Count != 2 {
		t.Fatalf("min 2 count = %d, want 2", p.Count)
	}
	mux := a.buildMux()
	get := func(path string) *httptest.ResponseRecorder {
		rec := httptest.NewRecorder()
		mux.ServeHTTP(rec, httptest.NewRequest(http.MethodGet, path, nil))
		return rec
	}
	rec := get("/patterns?min=2")
	var served []model.Pattern
	if rec.Code != http.StatusOK || json.Unmarshal(rec.Body.Bytes(), &served) != nil || len(served) != 1 {
		t.Fatalf("GET /patterns?min=2 = %d %s", rec.Code, rec.Body)
	}
	if rec := get("/patterns"); rec.Code != http.StatusOK || strings.TrimSpace(rec.Body.String()) != "[]" {
		t.Fatalf("GET /patterns = %d %q, want []", rec.Code, rec.Body)
	}
	for _, q := range []string{"min=1", "hours=0", "hours=721"} {
		if rec := get("/patterns?" + q); rec.Code != http.StatusBadRequest {
			t.Fatalf("GET /patterns?%s = %d, want 400", q, rec.Code)
		}
	}
}

// The disposition is the worst among OPEN flags: an acknowledged critical
// flag does not count.
func TestPatternDispositionWorstUnacked(t *testing.T) {
	a := explainTestAPI(t)
	now := time.Now()
	a.store.PutFlag(keychainFlag("crit", "codex", 7, now.Add(-3*time.Minute), 3))
	a.store.AcknowledgeFlag("crit")
	a.store.PutFlag(keychainFlag("warn", "codex", 7, now.Add(-2*time.Minute), 2))
	a.store.PutFlag(keychainFlag("benign", "codex", 7, now.Add(-time.Minute), 3))
	a.store.PutAdvisorVerdict("benign", "flag", model.AdvisorVerdict{Assessment: "benign", Confidence: 0.93, Rationale: "Own login.", CreatedAt: now})
	p := onlyPattern(t, a.computePatterns(now.Add(-time.Hour), 3))
	if p.Unacked != 2 || p.Disposition.State != model.DispositionWarning {
		t.Fatalf("unacked/disposition = %d/%+v, want 2/warning", p.Unacked, p.Disposition)
	}
	if p.FlagIDs[len(p.FlagIDs)-1] != "crit" {
		t.Fatalf("flag ids = %v, want the acknowledged one last", p.FlagIDs)
	}
}

func TestPatternAcknowledgedWhenAllAcked(t *testing.T) {
	a := explainTestAPI(t)
	now := time.Now()
	for i := 0; i < 3; i++ {
		id := fmt.Sprintf("d%d", i)
		a.store.PutFlag(keychainFlag(id, "codex", 7, now.Add(-time.Duration(i+1)*time.Minute), 3))
		a.store.AcknowledgeFlag(id)
	}
	p := onlyPattern(t, a.computePatterns(now.Add(-time.Hour), 3))
	if p.Unacked != 0 || p.Disposition.State != model.DispositionAcknowledged {
		t.Fatalf("unacked/disposition = %d/%+v, want 0/acknowledged", p.Unacked, p.Disposition)
	}
	if actionByID(p.Actions, "dismiss-all") != nil {
		t.Fatalf("actions = %+v, want no dismiss-all without open flags", p.Actions)
	}
}

// The summary names the agent, the count and the cadence; actions carry the
// served requests, the recommended one following the disposition.
func TestPatternSummaryAndActions(t *testing.T) {
	pinExplainHome(t, "/Users/x")
	a := explainTestAPI(t, AgentSummary{PID: 7, Name: "codex", StartedAt: "2026-09-23T10:00:00Z"})
	now := time.Now()
	for i := 0; i < 4; i++ {
		a.store.PutFlag(keychainFlag(fmt.Sprintf("c%d", i), "codex", 7, now.Add(-time.Duration(4-i)*12*time.Minute), 3))
	}
	for i := 0; i < 3; i++ {
		a.store.PutFlag(model.Flag{ID: fmt.Sprintf("e%d", i), Rule: "proxy-secret-leak", Severity: 3, TS: now.Add(-time.Duration(i+1) * time.Second), PID: 9, Agent: "claude",
			Evidence: []model.EvidenceItem{{Kind: "violation", Label: "proxy-secret-leak:aws-key"}, {Kind: "connect", Label: "api.example.com:443"}}})
		a.store.PutAdvisorVerdict(fmt.Sprintf("e%d", i), "flag", model.AdvisorVerdict{Assessment: "benign", Confidence: 0.9, Rationale: "Test key.", CreatedAt: now})
	}
	ps := a.computePatterns(now.Add(-24*time.Hour), 3)
	if len(ps) != 2 {
		t.Fatalf("patterns = %+v, want 2", ps)
	}
	byRule := map[string]model.Pattern{}
	for _, p := range ps {
		byRule[p.Rule] = p
	}
	kc := byRule["keychain-access"]
	for _, want := range []string{"codex touched the login keychain 4 times", "(1 process, 1 session)", "about every 12 minutes."} {
		if !strings.Contains(kc.Summary, want) {
			t.Fatalf("summary = %q, want %q", kc.Summary, want)
		}
	}
	if strings.Contains(kc.Summary, "c0") || kc.Cadence != "about every 12 minutes" {
		t.Fatalf("summary/cadence = %q/%q", kc.Summary, kc.Cadence)
	}
	mute, dismiss, kill := actionByID(kc.Actions, "mute-class"), actionByID(kc.Actions, "dismiss-all"), actionByID(kc.Actions, "kill")
	if mute == nil || mute.Body["rule"] != "keychain-access" || mute.Body["host"] != "*" {
		t.Fatalf("mute = %+v", mute)
	}
	if dismiss == nil || dismiss.Path != "/flags/acknowledge" || len(dismiss.Body["flag_ids"].([]string)) != 4 {
		t.Fatalf("dismiss-all = %+v", dismiss)
	}
	if kill == nil || !kill.Recommended || kill.Body["pid"] != int32(7) || kill.Body["started_at"] != "2026-09-23T10:00:00Z" {
		t.Fatalf("kill = %+v, want recommended for a critical pattern", kill)
	}

	eg := byRule["proxy-secret-leak"]
	if eg.Subject.Kind != "connect" || eg.Subject.Label != "api.example.com" || eg.Disposition.State != model.DispositionBenignLikely {
		t.Fatalf("egress subject/disposition = %+v/%+v", eg.Subject, eg.Disposition)
	}
	if !strings.Contains(eg.Summary, "claude sent a secret to api.example.com 3 times") || !strings.Contains(eg.Summary, "in bursts a few seconds apart") {
		t.Fatalf("egress summary = %q", eg.Summary)
	}
	allow := actionByID(eg.Actions, "allow-host")
	if allow == nil || !allow.Recommended || allow.Body["host"] != "api.example.com" || allow.Body["agent"] != "claude" {
		t.Fatalf("allow-host = %+v, want recommended for a benign-likely egress pattern", allow)
	}
	if actionByID(eg.Actions, "mute-rule-host") == nil || actionByID(eg.Actions, "kill") != nil {
		t.Fatalf("egress actions = %+v, want mute-rule-host and no kill (pid 9 is not live)", eg.Actions)
	}
}

// POST /flags/acknowledge takes a pattern's {flag_ids} in one call, bounded
// and validated, and still takes {flag_id}.
func TestAcknowledgeFlagIDs(t *testing.T) {
	a := explainTestAPI(t)
	for _, id := range []string{"f1", "f2", "f3", "f4"} {
		a.store.PutFlag(keychainFlag(id, "codex", 7, time.Now(), 2))
	}
	mux := a.buildMux()
	post := func(body string) *httptest.ResponseRecorder {
		rec := httptest.NewRecorder()
		mux.ServeHTTP(rec, httptest.NewRequest(http.MethodPost, "/flags/acknowledge", strings.NewReader(body)))
		return rec
	}
	if rec := post(`{"flag_ids":["f1","f2","f3"]}`); rec.Code != http.StatusOK || !strings.Contains(rec.Body.String(), `"count":3`) {
		t.Fatalf("flag_ids = %d %s", rec.Code, rec.Body)
	}
	for _, id := range []string{"f1", "f2", "f3"} {
		if f, _ := a.store.GetFlag(id); !f.Acknowledged {
			t.Fatalf("%s not acknowledged", id)
		}
	}
	ids := make([]string, 501)
	for i := range ids {
		ids[i] = fmt.Sprintf("x%d", i)
	}
	big, _ := json.Marshal(map[string]any{"flag_ids": ids})
	for _, body := range []string{string(big), `{"flag_ids":["f4","bad id"]}`, `{}`, `{"flag_ids":[]}`, `{"flag_id":"f4","flag_ids":["f4"]}`} {
		if rec := post(body); rec.Code != http.StatusBadRequest {
			t.Fatalf("body %.40s = %d, want 400", body, rec.Code)
		}
	}
	if f, _ := a.store.GetFlag("f4"); f.Acknowledged {
		t.Fatal("f4 acknowledged by a rejected request")
	}
	if rec := post(`{"flag_id":"f4"}`); rec.Code != http.StatusOK || !strings.Contains(rec.Body.String(), `"acknowledged":true`) {
		t.Fatalf("flag_id = %d %s", rec.Code, rec.Body)
	}
}

// Flags a pattern covers are one pattern item in the agent's group and the
// headline; their flag items are gone and groups and items still agree.
func TestAttentionPatternReplacesFlagItems(t *testing.T) {
	a := attentionAPI(t, []resource.Session{mkResourceSession(1, "codex", "/w")})
	now := time.Now()
	covered := map[string]bool{}
	for i := 0; i < 4; i++ {
		id := fmt.Sprintf("p%d", i)
		covered[id] = true
		a.store.PutFlag(keychainFlag(id, "codex", 1, now.Add(-time.Duration(i+1)*time.Second), 2))
	}
	a.store.PutFlag(model.Flag{ID: "tcc", Rule: "tcc-tamper", Severity: 3, TS: now, PID: 1, Agent: "codex"})

	p := a.computePosture()
	sum := 0
	var patternItems, flagItems []AttentionItem
	for _, g := range p.Groups {
		sum += len(g.Items)
		for _, it := range g.Items {
			switch it.Kind {
			case "pattern":
				patternItems = append(patternItems, it)
			case "flag":
				flagItems = append(flagItems, it)
			}
		}
	}
	if sum != p.NeedsYou || len(p.Items) != p.NeedsYou || p.NeedsYou != 2 {
		t.Fatalf("group items = %d, items = %d, needs_you = %d; want 2 each\nitems=%+v", sum, len(p.Items), p.NeedsYou, p.Items)
	}
	want := "codex|keychain-access|" + loginKeychain
	if len(patternItems) != 1 || patternItems[0].ID != want || patternItems[0].Count != 4 || patternItems[0].Rule != "keychain-access" ||
		patternItems[0].Disposition == nil || patternItems[0].Disposition.State != model.DispositionWarning || !strings.Contains(patternItems[0].Detail, "4 times") {
		t.Fatalf("pattern items = %+v", patternItems)
	}
	for _, it := range flagItems {
		if covered[it.ID] {
			t.Fatalf("flag item %s is covered by the pattern", it.ID)
		}
	}
	if len(flagItems) != 1 || flagItems[0].ID != "tcc" {
		t.Fatalf("flag items = %+v, want only tcc", flagItems)
	}
	headline := 0
	for _, it := range p.Items {
		if covered[it.ID] {
			t.Fatalf("headline item %+v is covered by the pattern", it)
		}
		if it.Kind == "pattern" && it.ID == want {
			headline++
		}
	}
	if headline != 1 {
		t.Fatalf("headline pattern items = %d, want 1\nitems=%+v", headline, p.Items)
	}
}
